Add more FilterParser and Matcher tests

Minoru opened the pull request:
These new tests expose different quirks in our existing C++ parser and matcher. Some Matcher tests actually check the behaviour of FilterParser; this is simply because Matcher's interface is way easier to use.

Reviews from everyone are welcome. Will merge in three days.

Minoru commented:
Since no-one reviewer this yet, I've allowed myself to force-push a rebase onto master, which resolves the conflict with test/matcher.cpp.

dennisschagt commented on test/filterparser.cpp:
```diff
@@ -64,3 +64,79 @@ TEST_CASE("Both = and == are accepted", "[FilterParser]")
+
+	// The following shouldn't pass, but it does. Further REQUIREs explain why:
+	// the NUL byte silently terminates parsing.
+	REQUIRE(fp.parse_string(std::string("attribute = \"hello\0world\"", 25)));
```
Is this something you want to get fixed?

Minoru replied:
Yes. It will be automatically fixed once we moved enough code to Rust: when config line is read straight into Rust `String` and then passed around without conversions to `char*`, internal NUL bytes won't be a problem anymore.

We could expedite this by passing strings' lengths over FFI, but I doubt it's worth the trouble. Never had a bug report from a user who wanted to use NUL in their filter expressions.

dennisschagt replied:
> Yes. It will be automatically fixed once we moved enough code to Rust

Sounds good

dennisschagt commented on test/matcher.cpp:
```diff
@@ -113,132 +226,420 @@ TEST_CASE("Matcher throws if regex passed to `=~` or `!~` is invalid",
 
-	m.parse("AAAA !~ \".\"");
-	REQUIRE_FALSE(m.matches(&mock));
+	SECTION("Works with strings") {
```
All tests in this block verify that `!~` returns false when the regular expression matches.
Might be nice to add a test in which the regelar expression does not match, resulting in `!~` returning true.

dennisschagt commented on test/matcher.cpp:
```diff
@@ -40,70 +40,183 @@ class MatcherMockMatchable : public Matchable {
+		MatcherMockMatchable mock({{"AAAA", "12345"}});
+
+		REQUIRE(m.parse("AAAA =~ 0:123456"));
+		REQUIRE_FALSE(m.matches(&mock));
```
Would this return true if the `mock` object was initialized with a value equal to `"0:123456"`?
e.g.: `MatcherMockMatchable mock({{"AAAA", "0:123456"}});`

Minoru replied:
That's a great catch! It does return `true`, because it treats everything as strings :(

Well, not for long. The new parser is stricter.

dennisschagt commented on test/matcher.cpp:
```diff
@@ -113,132 +226,420 @@ TEST_CASE("Matcher throws if regex passed to `=~` or `!~` is invalid",
-	m.parse("tags # \"foo\"");
-	REQUIRE(m.matches(&mock));
+	SECTION("Works with strings") {
+		MatcherMockMatchable mock({{"tags", "foo bar baz quux"}});
```
Given:
> "Operator `#` checks if space-separated list contains given value"

Might be nice to check if it checks for complete tags or also for parts of tags.

e.g. checking if `uu` matches (within `quux`):
```
REQUIRE(m.parse("tags # \"uu\""));
REQUIRE_FALSE(m.matches(&mock));
```

dennisschagt commented on test/matcher.cpp:
```diff
@@ -113,132 +226,420 @@ TEST_CASE("Matcher throws if regex passed to `=~` or `!~` is invalid",
+		REQUIRE_FALSE(m.matches(&mock));
+
+		REQUIRE(m.parse("AAAA between 0"));
+		REQUIRE_FALSE(m.matches(&mock));
```
Nitpick: `0` is not a range so these 2 lines fit better in the section called `"Doesn't work with numbers"`.

Minoru replied:
Oh, right. We already have "AAAA between 1" there, so I'll simply delete these two lines.

dennisschagt commented on test/matcher.cpp:
```diff
@@ -113,132 +226,420 @@ TEST_CASE("Matcher throws if regex passed to `=~` or `!~` is invalid",
+		SECTION("...converting numering prefix of the attribute if necessary") {
+			MatcherMockMatchable mock({{"value", "123four"}, {"practically_zero", "sure"}});
+
+			REQUIRE(m.parse("value between 0:124"));
```
Might be better to check with `"value between 122:124"` as it otherwise would pass if numeric prefix did not work and always returned 0.

Minoru commented:
Thanks for the review, @dennisschagt! I addressed all your comments in https://github.com/newsboat/newsboat/pull/853/commits/67c552775c1e1d04495b46bc1898d1cc9505542e, please re-review.

Minoru commented:
Thanks! Merged via https://github.com/newsboat/newsboat/commit/30f0aa49b9e13b572fc94fc8348ed883a3c682c0 (I rebased to get rid of the fixup commit)
